### PR TITLE
#635 - connectApp segment - user profile archiving

### PR DESCRIPTION
**Pull request by JoeArmani**
This PR handles user profile archiving in connectApp.
It also fixes a couple of lingering bugs in issue #20 - making user profiles editable in connectApp after verification.

issue 635: https://github.com/episphere/connect/issues/635
issue 20: https://github.com/episphere/connectApp/issues/20

-update & test logic around history writing
-added `refreshUserData()` to update userData after successful write. This gets past the issue where multiple writes in the same screen caused the latest write to overwrite the previous one.

Profile archiving rules:
-create new object {} in user history for each profile update and include timestamp for the user action.
-if field was empty and user adds data, write an empty string to history.
-if field was populated and user removes data, write an empty string to user profile.
-yes/no radio buttons (phone voicemail and text permissions) are handled when the associated phone number is changed.

Additional issue found: profile form was *sometimes* loading multiple times when navigating from other parts of the app.
Solution: Took more control over the build flow and added explicit calls to display the main form divs.

### CODE REVIEW

File: js/settingsHelpers.js [hunk 1]
```diff
@@ -94,15 +94,21 @@ export const hideOptionalElementsOnShowForm = elementsArray => {
...
-  document.getElementById('textPermissionNoRadio').checked = !canWeText;
-  document.getElementById('homeVoicemailPermissionYesRadio').checked = canWeVoicemailHome;
-  document.getElementById('homeVoicemailPermissionNoRadio').checked = !canWeVoicemailHome;
-  document.getElementById('otherVoicemailPermissionYesRadio').checked = canWeVoicemailOther;
-  document.getElementById('otherVoicemailPermissionNoRadio').checked = !canWeVoicemailOther;
+export const handleContactInformationRadioButtonPresets = (mobilePhoneNumberComplete, canWeVoicemailMobile, canWeText, homePhoneNumberComplete, canWeVoicemailHome, otherPhoneNumberComplete, canWeVoicemailOther) => {
+  if (mobilePhoneNumberComplete) {
+    document.getElementById('mobileVoicemailPermissionYesRadio').checked = canWeVoicemailMobile;
+    document.getElementById('mobileVoicemailPermissionNoRadio').checked = !canWeVoicemailMobile;
+    document.getElementById('textPermissionYesRadio').checked = canWeText;
```
Comment (anthonypetersen) [inline js/settingsHelpers.js:101]: Do HTML radio buttons not automatically un-select if another radio button in the same group is selected?

File: js/settingsHelpers.js [hunk 2]
```diff
@@ -426,12 +457,11 @@ const findChangedUserDataValues = (newUserData, existingUserData) => {
...
  * @param {string} type - the type of data being changed (e.g. name, contact info, mailing address, log-in email)
- * @param {*} isProfile
  * @returns {boolean} - whether the write operation was successful to control the UI messaging
  */
-const processUserDataUpdate = async (changedUserDataForProfile, changedUserDataForHistory, userHistory, type, isProfile) => {
+const processUserDataUpdate = async (changedUserDataForProfile, changedUserDataForHistory, userHistory, type) => {
   if (changedUserDataForProfile && Object.keys(changedUserDataForProfile).length !== 0) {
-    changedUserDataForProfile[cId.userProfileHistory] = updateUserHistory(changedUserDataForProfile, changedUserDataForHistory, userHistory, false);
+    changedUserDataForProfile[cId.userProfileHistory] = updateUserHistory(changedUserDataForHistory, userHistory);
     //console.log('changedUserData', changedUserDataForProfile);
```
Comment (anthonypetersen) [inline js/settingsHelpers.js:465]: remove console log if not needed

File: js/settingsHelpers.js (hunk 1 above)
Comment (JoeArmani) [reply js/settingsHelpers.js:101]: They sure do. This method pre-loads the previously selected values based on user profile data when a user opens the form to edit.

**Review by anthonypetersen — APPROVED** [review]
looks good, couple of comments left

File: js/settingsHelpers.js (hunk 2 above)
Comment (JoeArmani) [reply js/settingsHelpers.js:465]: Doing this now.

**Comment by JoeArmani** [comment]
-removed commented out `console.log()` Merging now.

File: js/settingsHelpers.js [hunk 3]
```diff
@@ -80,7 +80,7 @@ const updateElementContentAndDisplay = (element, text, content, display) => {
 };
 
 export const showAndPushElementToArrayIfExists = (value, element, isParentTruthy, array) => {
-  if (value !== null && value !== '' && element && isParentTruthy) {
+  if (value != null && value !== '' && element && isParentTruthy) {
```
[the remark this answers is not in the document]
Comment (JoeArmani) [reply js/settingsHelpers.js:83]: Yes! Oversight. Merged, but I'll mark this for a future update. Thanks

File: js/pages/settings.js [hunk 4]
```diff
@@ -138,14 +140,23 @@ export const renderSettingsPage = async () => {
...
+      loadMailingAddressElements();
+      loadSignInInformationElements();
+      showFormElements();
+      togglePendingVerificationMessage(userData);
   }
+  hideAnimation();
+};
+
+const showFormElements = () => {
+  document.getElementById('myProfileTextContainer').style.display = 'block';
```
[the remark this answers is not in the document]
Comment (JoeArmani) [reply js/pages/settings.js:155]: We sure could and this is a good suggestion. This is only called once on page load, I'm not convinced which way is more human-readable but it's a great one to consider.

**Pull request merged**